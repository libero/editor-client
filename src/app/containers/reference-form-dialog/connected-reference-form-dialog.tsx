import React, { useCallback } from 'react';
import { useDispatch } from 'react-redux';

import * as manuscriptActions from 'app/actions/manuscript.actions';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { ReferenceFormDialog } from 'app/containers/reference-form-dialog/reference-form-dialog';
import { Reference } from 'app/models/reference';

interface ConnectedReferenceFormDialogProps {
  reference?: Reference;
}

export const ConnectedReferenceFormDialog: React.FC<ConnectedReferenceFormDialogProps> = (props) => {
  const dispatch = useDispatch();
  const isNewReference = !props.reference;

  const closeDialog = useCallback(() => {
    dispatch(manuscriptEditorActions.hideModalDialog());
  }, [dispatch]);

  const handleAccept = useCallback(
    (reference: Reference) => {
      if (isNewReference) {
        dispatch(manuscriptActions.addReferenceAction(reference));
      } else {
        dispatch(manuscriptActions.updateReferenceAction(reference));
      }
      closeDialog();
    },
    [closeDialog, dispatch, isNewReference]
  );

  const handleDelete = useCallback(
    (reference: Reference) => {
      dispatch(manuscriptActions.deleteReferenceAction(reference));
      closeDialog();
    },
    [closeDialog, dispatch]
  );

  return (
    <ReferenceFormDialog
      reference={props.reference}
      onAccept={handleAccept}
      onDelete={handleDelete}
      onCancel={closeDialog}
    />
  );
};
