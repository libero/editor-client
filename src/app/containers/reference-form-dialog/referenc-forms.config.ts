import { ReferenceType } from 'app/models/reference';

export interface FormControlConfigType {
  type: string;
  label: string;
}

export const JOURNAL_FORM: Record<string, FormControlConfigType> = {
  year: { type: 'number', label: 'Year' },
  articleTitle: { type: 'rich-text', label: 'Title' },
  source: { type: 'rich-text', label: 'Journal title' },
  volume: { type: 'number', label: 'Volume' },
  firstPage: { type: 'string', label: 'First page' },
  lastPage: { type: 'string', label: 'Last page' },
  elocationId: { type: 'string', label: 'eLocation ID' },
  doi: { type: 'string', label: 'DOI' },
  pmid: { type: 'string', label: 'PMID' },
  inPress: { type: 'boolean', label: 'In press' }
};

export const PERIODICAL_FORM: Record<string, FormControlConfigType> = {
  date: { type: 'date', label: 'Date' },
  articleTitle: { type: 'rich-text', label: 'Title' },
  source: { type: 'rich-text', label: 'Periodical title' },
  volume: { type: 'number', label: 'Volume' },
  firstPage: { type: 'string', label: 'First page' },
  lastPage: { type: 'string', label: 'Last page' },
  extLink: { type: 'string', label: 'URL' }
};

export const BOOK_FORM: Record<string, FormControlConfigType> = {
  year: { type: 'number', label: 'Year' },
  chapterTitle: { type: 'rich-text', label: 'Chapter title' },
  source: { type: 'rich-text', label: 'Book title' },
  edition: { type: 'string', label: 'Edition' },
  volume: { type: 'number', label: 'Volume' },
  firstPage: { type: 'string', label: 'First page' },
  lastPage: { type: 'string', label: 'Last page' },
  elocationId: { type: 'string', label: 'eLocation ID' },
  publisherName: { type: 'string', label: 'Publisher Name' },
  publisherLocation: { type: 'string', label: 'Publisher Location' },
  editors: { type: 'editors-list', label: '' },
  doi: { type: 'string', label: 'DOI' },
  pmid: { type: 'string', label: 'PMID' },
  inPress: { type: 'boolean', label: 'In press' }
};

export const REPORT_FORM: Record<string, FormControlConfigType> = {
  year: { type: 'number', label: 'Year' },
  source: { type: 'rich-text', label: 'Title' },
  publisherName: { type: 'string', label: 'Publisher Name' },
  doi: { type: 'string', label: 'DOI' },
  extLink: { type: 'string', label: 'URL' }
};

export const DATA_FORM: Record<string, FormControlConfigType> = {
  year: { type: 'number', label: 'Year' },
  dataTitle: { type: 'rich-text', label: 'Data title' },
  source: { type: 'rich-text', label: 'Source' },
  version: { type: 'string', label: 'Version' },
  doi: { type: 'string', label: 'DOI' },
  accessionId: { type: 'string', label: 'Accession' },
  accessionUrl: { type: 'string', label: 'Accession URL' },
  extLink: { type: 'string', label: 'URL' },
  specificUse: { type: 'string', label: 'Specific Use' }
};

export const WEB_FORM: Record<string, FormControlConfigType> = {
  dateInCitation: { type: 'date', label: 'Date' },
  year: { type: 'number', label: 'Year' },
  articleTitle: { type: 'rich-text', label: 'Article title' },
  source: { type: 'rich-text', label: 'Source' },
  extLink: { type: 'string', label: 'URL' }
};

export const PREPRINT_FORM: Record<string, FormControlConfigType> = {
  year: { type: 'number', label: 'Year' },
  dateInCitation: { type: 'date', label: 'Date' },
  articleTitle: { type: 'rich-text', label: 'Article title' },
  source: { type: 'rich-text', label: 'Journal title' },
  doi: { type: 'string', label: 'DOI' },
  pmid: { type: 'string', label: 'PMID' },
  extLink: { type: 'string', label: 'URL' }
};

export const SOFTWARE_FORM: Record<string, FormControlConfigType> = {
  year: { type: 'number', label: 'Year' },
  dataTitle: { type: 'rich-text', label: 'Title' },
  source: { type: 'rich-text', label: 'Source' },
  publisherName: { type: 'string', label: 'Publisher Name' },
  publisherLocation: { type: 'string', label: 'Publisher Location' },
  version: { type: 'string', label: 'Version' },
  doi: { type: 'string', label: 'DOI' },
  pmid: { type: 'string', label: 'PMID' },
  extLink: { type: 'string', label: 'ext link' }
};

export const CONFPROC_FORM: Record<string, FormControlConfigType> = {
  year: { type: 'number', label: 'Year' },
  articleTitle: { type: 'rich-text', label: 'Article title' },
  conferenceName: { type: 'string', label: 'Conference name' },
  conferenceLocation: { type: 'string', label: 'Conference location' },
  conferenceDate: { type: 'date', label: 'Conference date' },
  volume: { type: 'number', label: 'Volume' },
  firstPage: { type: 'string', label: 'First page' },
  lastPage: { type: 'string', label: 'Last page' },
  elocationId: { type: 'string', label: 'eLocation ID' },
  doi: { type: 'string', label: 'DOI' },
  pmid: { type: 'string', label: 'PMID' },
  extLink: { type: 'string', label: 'ext link' }
};

export const THESIS_FORM: Record<string, FormControlConfigType> = {
  year: { type: 'number', label: 'Year' },
  articleTitle: { type: 'rich-text', label: 'Article title' },
  publisherName: { type: 'string', label: 'Publisher Name' },
  publisherLocation: { type: 'string', label: 'Publisher Location' },
  doi: { type: 'string', label: 'DOI' },
  pmid: { type: 'string', label: 'PMID' },
  extLink: { type: 'string', label: 'ext link' }
};

export const PATENT_FORM: Record<string, FormControlConfigType> = {
  year: { type: 'number', label: 'Year' },
  articleTitle: { type: 'rich-text', label: 'Article title' },
  source: { type: 'rich-text', label: 'Source' },
  publisherName: { type: 'string', label: 'Publisher Name' },
  patent: { type: 'string', label: 'Patent' },
  doi: { type: 'string', label: 'DOI' },
  pmid: { type: 'string', label: 'PMID' },
  extLink: { type: 'string', label: 'ext link' }
};

export function getFormConfigForType(type: ReferenceType): Record<string, FormControlConfigType> {
  return {
    journal: JOURNAL_FORM,
    periodical: PERIODICAL_FORM,
    book: BOOK_FORM,
    report: REPORT_FORM,
    data: DATA_FORM,
    web: WEB_FORM,
    preprint: PREPRINT_FORM,
    software: SOFTWARE_FORM,
    confproc: CONFPROC_FORM,
    thesis: THESIS_FORM,
    patent: PATENT_FORM
  }[type];
}
