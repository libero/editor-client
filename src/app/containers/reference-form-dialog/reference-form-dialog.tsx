import React, { useCallback, useState, ChangeEvent } from 'react';
import { get, has, pick, set } from 'lodash';
import DeleteIcon from '@material-ui/icons/Delete';
import classNames from 'classnames';
import { IconButton } from '@material-ui/core';
import { EditorState } from 'prosemirror-state';
import { Alert } from '@material-ui/lab';

import { Reference, ReferenceContributor, ReferenceInfoType } from '../../models/reference';
import { Select } from '../../components/select';
import { useReferenceFormStyles } from './styles';
import { ActionButton } from '../../components/action-button';
import { renderConfirmDialog } from '../../components/prompt-dialog';
import { ReferenceContributorsList } from './reference-contributors-list';
import { FormControlConfigType, getFormConfigForType } from './reference-forms.config';
import { renderFormControl } from './reference-form-renderer';
import refFormGrid from '../../styles/form-grid.module.scss';
import { objectsEqual } from '../../utils/view.utils';
import { ReferenceType } from '../../models/reference-type';

interface ReferenceFormDialogProps {
  reference?: Reference;
  onAccept(reference: Reference): void;
  onCancel(): void;
  onDelete(reference: Reference): void;
}

export const ReferenceFormDialog: React.FC<ReferenceFormDialogProps> = ({
  reference,
  onAccept,
  onCancel,
  onDelete
}) => {
  const classes = useReferenceFormStyles();
  const [isConfirmShown, setConfirmShow] = useState<boolean>(false);
  const isNewReference = !reference;
  const [userReference, setReference] = useState<Reference>(reference || new Reference());
  const [missingFieldsInfo, setMissingFieldsInfo] = useState<Partial<ReferenceInfoType>>({});
  const [missingFieldsConfig, setMissingFieldsConfig] = useState<Record<string, FormControlConfigType>>({});
  const handleReferenceTypeChange = useCallback(
    (event: ChangeEvent<{ name: string; value: ReferenceType }>) => {
      const newRef = userReference.clone();
      newRef.type = event.target['value'];
      newRef.referenceInfo = transferValues(userReference.referenceInfo, newRef.referenceInfo, missingFieldsInfo);
      if (userReference.referenceInfo) {
        const missingInfo = getDiffFieldValues(userReference.referenceInfo, newRef.referenceInfo, missingFieldsInfo);
        setMissingFieldsInfo(missingInfo);
        setMissingFieldsConfig(getDiffFieldsConfig(userReference.type, newRef.type, missingFieldsConfig, missingInfo));
      }
      setReference(newRef);
    },
    [userReference, missingFieldsInfo, missingFieldsConfig]
  );

  const handleAuthorsListChange = useCallback(
    (refAuthors: ReferenceContributor[]) => {
      const newRef = userReference.clone();
      newRef.authors = refAuthors;
      setReference(newRef);
    },
    [userReference, setReference]
  );

  const handleRefInfoChange = useCallback(
    (name, value) => {
      const newRef = userReference.clone();
      set(newRef.referenceInfo, name, value);
      setReference(newRef);
    },
    [userReference, setReference]
  );

  const handleMissingFieldsInfoChange = useCallback(
    (name, value) => {
      setMissingFieldsInfo({
        ...missingFieldsInfo,
        [name]: value
      });
    },
    [missingFieldsInfo]
  );

  const handleDone = useCallback(() => {
    userReference.authors = userReference.authors.filter(
      (author) => author['firstName'] || author['lastName'] || author['groupName']
    );
    if (!objectsEqual(userReference, reference)) {
      onAccept(userReference);
    } else {
      onCancel();
    }
  }, [userReference, reference, onAccept, onCancel]);

  const handleReject = useCallback(() => {
    setConfirmShow(false);
  }, [setConfirmShow]);

  const handleDeleteRow = useCallback(
    (fieldName: string) => () => {
      const updateMissingFields = { ...missingFieldsInfo };
      delete updateMissingFields[fieldName];
      setMissingFieldsInfo(updateMissingFields);
    },
    [missingFieldsInfo]
  );

  const handleDelete = useCallback(() => {
    setConfirmShow(true);
  }, [setConfirmShow]);

  const handleAccept = useCallback(() => {
    setConfirmShow(false);
    onDelete(userReference);
  }, [userReference, onDelete]);

  const formConfig = getFormConfigForType(userReference.type);
  const form = formConfig
    ? Object.entries(formConfig).map(([key, config]) => {
        return renderFormControl(
          config.type,
          config.label,
          key,
          classNames(config.className),
          get(userReference, `referenceInfo.${key}`),
          handleRefInfoChange
        );
      })
    : undefined;

  const diffForm = Object.entries(missingFieldsInfo).map(([key, value]) => {
    const config = missingFieldsConfig[key];
    return (
      <div className={classNames(classes.missingFieldsRow, classes.inputField)}>
        {renderFormControl(
          config.type,
          config.label,
          key,
          classes.missingField,
          value,
          handleMissingFieldsInfoChange,
          true
        )}
        <IconButton classes={{ root: classes.deleteButton }} onClick={handleDeleteRow(key)}>
          <DeleteIcon fontSize="small" />
        </IconButton>
      </div>
    );
  });

  return (
    <section className={classes.root}>
      <Select
        className={classes.inputField}
        name="type"
        test-id={'ref-type'}
        placeholder="Please select"
        fullWidth
        blankValue={undefined}
        label="Reference type"
        value={userReference.type}
        onChange={handleReferenceTypeChange}
        options={[
          { label: 'Journal Article', value: 'journal' },
          { label: 'Book', value: 'book' },
          { label: 'Data', value: 'data' },
          { label: 'Software', value: 'software' },
          { label: 'Preprint', value: 'preprint' },
          { label: 'Web Article', value: 'web' },
          { label: 'Conference proceedings', value: 'confproc' },
          { label: 'Report', value: 'report' },
          { label: 'Thesis', value: 'thesis' },
          { label: 'Patent', value: 'patent' },
          { label: 'Periodical Article', value: 'periodical' }
        ]}
      />
      <ReferenceContributorsList
        className={classes.inputField}
        addCtaLabel={'Add Author'}
        entityName={'author'}
        test-id={'ref-authors'}
        refContributors={userReference.authors}
        onChange={handleAuthorsListChange}
      />
      <div className={classNames(classes.inputField, refFormGrid.container)}>{form}</div>
      {diffForm.length > 0 ? (
        <div className={classes.missingFieldsSection}>
          <Alert elevation={0} variant="filled" severity="error" classes={{ root: classes.missingFieldsMessage }}>
            Unsupported elements for reference type
          </Alert>
          {diffForm}
        </div>
      ) : undefined}
      <div className={classes.buttonPanel}>
        {!isNewReference ? <ActionButton variant="outlinedWarning" onClick={handleDelete} title="Delete" /> : undefined}
        <div aria-hidden={true} className={classes.spacer}></div>
        <ActionButton variant="secondaryOutlined" onClick={onCancel} title="Cancel" />
        <ActionButton variant="primaryContained" onClick={handleDone} title="Done" />
      </div>
      {isConfirmShown
        ? renderConfirmDialog(
            'You are deleting a reference',
            'Deleting a reference will remove all linked citations from this article. Are you sure you want to proceed?',
            handleAccept,
            handleReject
          )
        : undefined}
    </section>
  );
};

function transferValues(
  prevRefInfo: ReferenceInfoType,
  nextRefInfo: ReferenceInfoType,
  prevMissingInfo: Partial<ReferenceInfoType> = {}
): ReferenceInfoType {
  Object.entries(nextRefInfo).forEach(([key, value]) => {
    nextRefInfo[key] = get(prevRefInfo, key, get(prevMissingInfo, key, value));
  });

  return nextRefInfo;
}

function getDiffFieldValues(
  prevRefInfo: ReferenceInfoType,
  nextRefInfo: ReferenceInfoType,
  prevMissingInfo: Partial<ReferenceInfoType> = {}
): Partial<ReferenceInfoType> {
  const diffKeys = Object.keys(prevRefInfo).filter((key) => !has(nextRefInfo, key) && !isValueEmpty(prevRefInfo[key]));
  const unsalvagedFieldKeys = Object.keys(prevMissingInfo).filter((key) => !has(nextRefInfo, key));
  return {
    ...pick(prevRefInfo, diffKeys),
    ...pick(prevMissingInfo, unsalvagedFieldKeys)
  };
}

function getDiffFieldsConfig(
  prevType: ReferenceType,
  nextType: ReferenceType,
  prevMissingFieldsConfig: Record<string, FormControlConfigType> = {},
  currentMissingFields: Partial<ReferenceInfoType> = {}
): Record<string, FormControlConfigType> {
  const prevConfig = getFormConfigForType(prevType);
  const nextConfig = getFormConfigForType(nextType);
  const diffKeys = Object.keys(prevConfig).filter((key: ReferenceType) => !Boolean(nextConfig[key]));
  const unsalvagedFieldsConfigKeys = Object.keys(currentMissingFields).filter((key) =>
    has(prevMissingFieldsConfig, key)
  );
  return {
    ...pick(prevConfig, diffKeys),
    ...pick(prevMissingFieldsConfig, unsalvagedFieldsConfigKeys)
  };
}

function isValueEmpty(value: EditorState | Array<ReferenceContributor> | string | number): boolean {
  if (value instanceof EditorState) {
    return !value.doc || value.doc.childCount === 0 || value.doc.textContent.trim().length === 0;
  }

  if (Array.isArray(value)) {
    return value.length === 0;
  }

  return !value;
}
