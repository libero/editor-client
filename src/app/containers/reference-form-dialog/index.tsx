import React, { useCallback, useState, ChangeEvent } from 'react';
import { useDispatch } from 'react-redux';

import { createBlankReference, Reference, ReferencePerson, ReferenceType } from 'app/models/reference';
import { Select } from 'app/components/select';
import { useReferenceFormStyles } from 'app/containers/reference-form-dialog/styles';
import { ActionButton } from 'app/components/action-button';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import * as manuscriptActions from 'app/actions/manuscript.actions';
import { renderConfirmDialog } from 'app/components/prompt-dialog';
import { ReferenceAuthorsList } from 'app/containers/reference-form-dialog/reference-authors-list';

interface ReferenceFormDialogProps {
  reference?: Reference;
}

export const ReferenceFormDialog: React.FC<ReferenceFormDialogProps> = ({ reference }) => {
  const classes = useReferenceFormStyles();
  const [isConfirmShown, setConfirmSnow] = useState<boolean>(false);
  const isNewReference = !reference;
  const [userReference, setReference] = useState<Reference>(reference || createBlankReference());
  const dispatch = useDispatch();

  const handleReferenceTypeChange = useCallback(
    (event: ChangeEvent<{ name: string; value: ReferenceType }>) => {
      setReference({
        ...userReference,
        type: event.target['value']
      });
    },
    [userReference]
  );

  const closeDialog = useCallback(() => {
    dispatch(manuscriptEditorActions.hideModalDialog());
  }, [dispatch]);

  const handleAuthorsListChange = useCallback(
    (refAuthors: ReferencePerson[]) => {
      setReference({
        ...userReference,
        authors: refAuthors
      });
    },
    [userReference, setReference]
  );

  const handleDone = useCallback(() => {
    userReference.authors = userReference.authors.filter(
      (author) => author['firstName'] || author['lastName'] || author['groupName']
    );
    if (isNewReference) {
      dispatch(manuscriptActions.addReferenceAction(userReference));
    } else {
      dispatch(manuscriptActions.updateReferenceAction(userReference));
    }
    closeDialog();
  }, [userReference, closeDialog, dispatch, isNewReference]);

  const handleReject = useCallback(() => {
    setConfirmSnow(false);
  }, [setConfirmSnow]);

  const handleDelete = useCallback(() => {
    setConfirmSnow(true);
  }, [setConfirmSnow]);

  const handleAccept = useCallback(() => {
    setConfirmSnow(false);
    dispatch(manuscriptActions.deleteReferenceAction(userReference));
    closeDialog();
  }, [closeDialog, dispatch, userReference]);

  return (
    <section className={classes.root}>
      <Select
        className={classes.inputField}
        name="type"
        test-id={'ref-type'}
        placeholder="Please select"
        fullWidth
        blankValue={undefined}
        label="Reference type"
        value={userReference.type}
        onChange={handleReferenceTypeChange}
        options={[
          { label: 'Journal Article', value: 'journal' },
          { label: 'Periodical Article', value: 'periodical' },
          { label: 'Book', value: 'book' },
          { label: 'Report', value: 'report' },
          { label: 'Data', value: 'data' },
          { label: 'Web Article', value: 'web' },
          { label: 'Pre-print Article', value: 'preprint' },
          { label: 'Software', value: 'software' },
          { label: 'Conference proceedings', value: 'confproc' },
          { label: 'Thesis', value: 'thesis' },
          { label: 'Patent', value: 'patent' }
        ]}
      />
      <ReferenceAuthorsList
        test-id={'ref-authors'}
        refAuthors={userReference.authors}
        onChange={handleAuthorsListChange}
      />
      <div className={classes.buttonPanel}>
        {!isNewReference ? <ActionButton variant="outlinedWarning" onClick={handleDelete} title="Delete" /> : undefined}
        <div aria-hidden={true} className={classes.spacer}></div>
        <ActionButton variant="secondaryOutlined" onClick={closeDialog} title="Cancel" />
        <ActionButton variant="primaryContained" onClick={handleDone} title="Done" />
      </div>
      {isConfirmShown
        ? renderConfirmDialog(
            'You are deleting a reference',
            'Deleting a a reference can leave unlinked citations. Are you sure you want to proceed?',
            handleAccept,
            handleReject
          )
        : undefined}
    </section>
  );
};
