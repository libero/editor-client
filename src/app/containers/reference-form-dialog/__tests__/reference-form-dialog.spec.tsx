import React from 'react';
import configureMockStore from 'redux-mock-store';
import { create } from 'react-test-renderer';
import { Provider } from 'react-redux';

import { givenState } from 'app/test-utils/reducer-test-helpers';
import { mount } from 'enzyme';
import * as manuscriptActions from 'app/actions/manuscript.actions';
import { PromptDialog } from 'app/components/prompt-dialog';
import { ConnectedReferenceFormDialog } from 'app/containers/reference-form-dialog/connected-reference-form-dialog';
import { createReferenceAnnotatedValue, ReferenceType } from 'app/models/reference-type';
import { Reference } from 'app/models/reference';
import { JSONObject } from 'app/types/utility.types';

jest.mock('@material-ui/core', () => ({
  Dialog: () => <div data-cmp="Dialog"></div>,
  Menu: ({ children }) => <div data-cmp="Menu">{children}</div>,
  Select: ({ children }) => <div data-cmp="Index">{children}</div>,
  MenuItem: ({ children }) => <div data-cmp="MenuItem">{children}</div>,
  FormControl: ({ children }) => <div data-cmp="FormControl">{children}</div>,
  FormControlLabel: ({ children }) => <div data-cmp="FormControlLabel">{children}</div>,
  InputLabel: ({ children }) => <div data-cmp="InputLabel">{children}</div>,
  TextField: ({ children }) => <div data-cmp="TextField">{children}</div>,
  Checkbox: ({ children }) => <div data-cmp="CheckboxInput">{children}</div>,
  IconButton: ({ children }) => <div data-cmp="iconButton">{children}</div>,
  Button: ({ children }) => <div data-cmp="Button">{children}</div>
}));

jest.mock('app/components/rich-text-input', () => ({
  RichTextInput: ({ children }) => <div data-cmp="RichTextInput">{children}</div>
}));

jest.mock('@material-ui/lab', () => ({
  ToggleButtonGroup: ({ children }) => <div data-cmp="ToggleButtonGroup">{children}</div>,
  ToggleButton: ({ children }) => <div data-cmp="ToggleButton">{children}</div>
}));

jest.mock('@material-ui/core/styles', () => {
  return {
    ThemeProvider: ({ children }) => <>{children}</>,
    createMuiTheme: jest.fn(),
    makeStyles: jest.requireActual('@material-ui/core/styles').makeStyles
  };
});

describe('Reference form dialog', () => {
  const mockStore = configureMockStore([]);
  let mockState;

  beforeEach(() => {
    mockState = givenState({
      references: [
        new Reference({
          _id: 'some_id',
          _type: 'journal' as ReferenceType,
          authors: [{ firstName: 'First Name', lastName: 'Last Name' }, { groupName: 'Group Name' }],
          referenceInfo: {
            year: '2012',
            source: stringToEditorState('Science'),
            articleTitle: stringToEditorState(
              'Molecular architecture and assembly principles of <italic>Vibrio cholerae</italic> biofilms'
            ),
            doi: '',
            pmid: '',
            elocationId: '',
            firstPage: '236',
            lastPage: '239',
            inPress: false,
            volume: '337'
          }
        })
      ]
    });
  });

  it('renders form', () => {
    const store = mockStore({ manuscript: mockState });
    const wrapper = create(
      <Provider store={store}>
        <ConnectedReferenceFormDialog reference={mockState.data.present.references[0]} />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should not dispatch update action when no changes', () => {
    const store = mockStore({ manuscript: mockState });
    jest.spyOn(store, 'dispatch');

    const wrapper = mount(
      <Provider store={store}>
        <ConnectedReferenceFormDialog reference={mockState.data.present.references[0]} />
      </Provider>
    );

    wrapper.find({ title: 'Done' }).prop('onClick')();
    expect(store.dispatch).not.toHaveBeenCalledWith(
      manuscriptActions.updateReferenceAction(mockState.data.present.references[0])
    );
  });

  it('should dispatch update action', () => {
    const store = mockStore({ manuscript: mockState });
    jest.spyOn(store, 'dispatch');

    const wrapper = mount(
      <Provider store={store}>
        <ConnectedReferenceFormDialog reference={mockState.data.present.references[0]} />
      </Provider>
    );
    wrapper.find({ name: 'firstName' }).at(0).prop('onChange')({ target: { name: 'firstName', value: 'new value' } });
    const updatedReference = mockState.data.present.references[0].clone();
    updatedReference.authors[0].firstName = 'new value';

    wrapper.update();
    wrapper.find({ title: 'Done' }).prop('onClick')();
    wrapper.update();
    expect(store.dispatch).toHaveBeenCalledWith(manuscriptActions.updateReferenceAction(updatedReference));
  });

  it('should dispatch add action', () => {
    const store = mockStore({ manuscript: mockState });
    jest.spyOn(store, 'dispatch');

    const wrapper = mount(
      <Provider store={store}>
        <ConnectedReferenceFormDialog />
      </Provider>
    );

    wrapper.find({ 'test-id': 'ref-authors' }).prop('onChange')([{ firstName: 'First Name', lastName: 'last name' }]);
    wrapper.update();
    wrapper.find({ 'test-id': 'ref-type' }).prop('onChange')({ target: { value: 'journal' } });
    wrapper.update();
    wrapper.find({ title: 'Done' }).prop('onClick')();

    expect(store.dispatch).toHaveBeenCalledWith(manuscriptActions.addReferenceAction(expect.any(Reference)));
  });

  it('should dispatch delete action', () => {
    const store = mockStore({ manuscript: mockState });
    jest.spyOn(store, 'dispatch');

    const wrapper = mount(
      <Provider store={store}>
        <ConnectedReferenceFormDialog reference={mockState.data.present.references[0]} />
      </Provider>
    );

    wrapper.find({ title: 'Delete' }).prop('onClick')();
    wrapper.update();
    wrapper.find(PromptDialog).prop('onAccept')();

    expect(store.dispatch).toHaveBeenCalledWith(
      manuscriptActions.deleteReferenceAction(mockState.data.present.references[0])
    );
  });

  function stringToEditorState(xmlContent: string): JSONObject {
    const el = document.createElement('div');
    el.innerHTML = xmlContent;
    return createReferenceAnnotatedValue(el).toJSON();
  }
});
