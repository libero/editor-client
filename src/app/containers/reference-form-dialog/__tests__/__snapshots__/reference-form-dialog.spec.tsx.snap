// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reference form dialog renders form 1`] = `
<section
  className="makeStyles-root-1"
>
  <div
    data-cmp="FormControl"
  >
    <div
      data-cmp="InputLabel"
    >
      Reference type
    </div>
    <div
      data-cmp="Index"
    >
      <div
        data-cmp="MenuItem"
      >
        Please select
      </div>
      <div
        data-cmp="MenuItem"
      >
        Journal Article
      </div>
      <div
        data-cmp="MenuItem"
      >
        Book
      </div>
      <div
        data-cmp="MenuItem"
      >
        Data
      </div>
      <div
        data-cmp="MenuItem"
      >
        Software
      </div>
      <div
        data-cmp="MenuItem"
      >
        Preprint
      </div>
      <div
        data-cmp="MenuItem"
      >
        Web Article
      </div>
      <div
        data-cmp="MenuItem"
      >
        Conference proceedings
      </div>
      <div
        data-cmp="MenuItem"
      >
        Report
      </div>
      <div
        data-cmp="MenuItem"
      >
        Thesis
      </div>
      <div
        data-cmp="MenuItem"
      >
        Patent
      </div>
      <div
        data-cmp="MenuItem"
      >
        Periodical Article
      </div>
    </div>
  </div>
  <section
    className="makeStyles-inputField-2"
  >
     
    <section
      className="makeStyles-contributorInputFields-12"
    >
      <img
        alt="drag handle"
        aria-hidden={true}
        className="drag-handle"
        src="drag-indicator-grey.svg"
      />
      <div
        data-cmp="TextField"
      />
      <div
        data-cmp="TextField"
      />
      <div
        data-cmp="iconButton"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
          />
        </svg>
      </div>
      <div
        data-cmp="Menu"
      >
        <div
          data-cmp="MenuItem"
        >
          Set to group author
        </div>
        <div
          data-cmp="MenuItem"
        >
          Delete 
          author
        </div>
      </div>
    </section>
    <section
      className="makeStyles-contributorInputFields-12"
    >
      <img
        alt="drag handle"
        aria-hidden={true}
        className="drag-handle"
        src="drag-indicator-grey.svg"
      />
      <div
        data-cmp="TextField"
      />
      <div
        data-cmp="iconButton"
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
          />
        </svg>
      </div>
      <div
        data-cmp="Menu"
      >
        <div
          data-cmp="MenuItem"
        >
          Set to individual author
        </div>
        <div
          data-cmp="MenuItem"
        >
          Delete 
          author
        </div>
      </div>
    </section>
    <div
      data-cmp="Button"
    >
      Add Author
    </div>
     
  </section>
  <div
    className="makeStyles-inputField-2 container"
  >
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="RichTextInput"
    />
    <div
      data-cmp="RichTextInput"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="TextField"
    />
    <div
      data-cmp="FormControlLabel"
    />
  </div>
  <div
    className="makeStyles-buttonPanel-8"
  >
    <div
      data-cmp="Button"
    >
      Delete
    </div>
    <div
      aria-hidden={true}
      className="makeStyles-spacer-9"
    />
    <div
      data-cmp="Button"
    >
      Cancel
    </div>
    <div
      data-cmp="Button"
    >
      Done
    </div>
  </div>
</section>
`;
